src/goto-instrument/code_contracts.cpp

```diff
@@ -209,14 +253,46 @@ void code_contractst::check_apply_invariants(
     converter.goto_convert(assertion, havoc_code, mode);
     havoc_code.instructions.back().source_location.set_comment(
       "Check that loop invariant is preserved");
-    auto offset = havoc_code.instructions.size();
-    goto_function.body.insert_before_swap(loop_end, havoc_code);
-    std::advance(loop_end, offset);
   }
 
+  if(!decreases_clause.is_nil())
+  {
+    // Generate: a declaration of a temporary variable that stores the decreases
+    // clause after one arbitrary iteration of the loop
+    havoc_code.add(goto_programt::make_decl(
+      new_decreases_symbol, loop_head->source_location));
+
+    // Generate: an assignment to store the decreases clause's value after one
+    // iteration of the loop
+    code_assignt new_decreases_assignment{new_decreases_symbol,
+                                          decreases_clause};
+    new_decreases_assignment.add_source_location() = loop_head->source_location;
+    converter.goto_convert(new_decreases_assignment, havoc_code, mode);
+
+    // Generate: assertion that the decreases clause's value after the loop is
+    // smaller than the value before the loop
+    code_assertt monotonic_decreasing_assertion{
+      binary_relation_exprt(new_decreases_symbol, ID_lt, old_decreases_symbol)};
+    monotonic_decreasing_assertion.add_source_location() =
+      loop_head->source_location;
+    converter.goto_convert(monotonic_decreasing_assertion, havoc_code, mode);
+    havoc_code.instructions.back().source_location.set_comment(
+      "Check monotonicity of decreases clause");
```

Comment:
"monotonicity" is perhaps not the right word. It might not be an intuitive description for a broader audience, and being pedantic, you need strict monotonicity to get termination.
How about "ranking function decreases"?

Reply:
I think "ranking function" might also sound unfamiliar, plus we haven't used ranking function elsewhere (the annotation just reads `__CPROVER_decreases`).

How about simply, `"Check decreases clause on loop iteration"`?

Reply:
Thanks a lot for both of your suggestions. I will change the comment to `"Check decreases clause on loop iteration"`.